File: javascript/jssearch/src/filetypes.js
/*
 * filetypes.js
 *
 * identifies file types (archive, binary, text, unknown)
 */

var common = require('./common.js');
var config = require('./config.js');
var FileType = require('./filetype.js').FileType;
var FileUtil = require('./fileutil.js').FileUtil;

function FileTypes() {
    "use strict";
    let self = this;

    const fileTypeMap = (() => {
        let fs = require('fs');

        let json = '';
        if (fs.existsSync(FileUtil.expandPath(config.FILETYPESJSONPATH))) {
            json = fs.readFileSync(FileUtil.expandPath(config.FILETYPESJSONPATH)).toString();
        } else {
            throw new Error('File not found: ' + config.FILETYPESJSONPATH);
        }

        let fileTypeMap = {};

        let obj = JSON.parse(json);
        if (obj.hasOwnProperty('filetypes') && Array.isArray(obj['filetypes'])) {
            obj['filetypes'].forEach(ft => {
                let typename = ft['type'];
                let extensions = ft['extensions'];
                fileTypeMap[typename] = common.setFromArray(extensions);
            });
        } else throw new Error("Invalid filetypes file: " + config.FILETYPESJSONPATH);

        fileTypeMap.text = [].concat(fileTypeMap.text, fileTypeMap.code, fileTypeMap.xml);
        fileTypeMap.searchable = [].concat(fileTypeMap.text, fileTypeMap.binary,
            fileTypeMap.archive);

        return fileTypeMap;
    })();

    this.getFileType = function (filename) {
        if (self.isCodeFile(filename))
            return FileType.CODE;
        if (self.isXmlFile(filename))
            return FileType.XML;
        if (self.isTextFile(filename))
            return FileType.TEXT;
        if (self.isBinaryFile(filename))
            return FileType.BINARY;
        if (self.isArchiveFile(filename))
            return FileType.ARCHIVE;
        return FileType.UNKNOWN;
    };

    this.getFileTypeAsync = function (filename, cb) {
        if (self.isCodeFile(filename))
            return cb(FileType.CODE);
        if (self.isXmlFile(filename))
            return cb(FileType.XML);
        if (self.isTextFile(filename))
            return cb(FileType.TEXT);
        if (self.isBinaryFile(filename))
            return cb(FileType.BINARY);
        if (self.isArchiveFile(filename))
            return cb(FileType.ARCHIVE);
        cb(FileType.UNKNOWN);
    };

    this.isArchiveFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.archive.indexOf(ext) > -1;
    };

    this.isBinaryFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.binary.indexOf(ext) > -1;
    };

    this.isCodeFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.code.indexOf(ext) > -1;
    };

    this.isSearchableFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.searchable.indexOf(ext) > -1;
    };

    this.isTextFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.text.indexOf(ext) > -1;
    };

    this.isXmlFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.xml.indexOf(ext) > -1;
    };

    this.isUnknownFile = function (filename) {
        let ext = FileUtil.getExtension(filename);
        return fileTypeMap.searchable.indexOf(ext) === -1;
    };
}

FileTypes.fromName = function (name) {
    if (name.toUpperCase() === 'TEXT')
        return FileType.TEXT;
    if (name.toUpperCase() === 'BINARY')
        return FileType.BINARY;
    if (name.toUpperCase() === 'ARCHIVE')
        return FileType.ARCHIVE;
    if (name.toUpperCase() === 'CODE')
        return FileType.CODE;
    if (name.toUpperCase() === 'XML')
        return FileType.XML;
    return FileType.UNKNOWN;
};

FileTypes.toName = function (fileType) {
    if (fileType === FileType.ARCHIVE)
        return 'ARCHIVE';
    if (fileType === FileType.BINARY)
        return 'BINARY';
    if (fileType === FileType.CODE)
        return 'CODE';
    if (fileType === FileType.TEXT)
        return 'TEXT';
    if (fileType === FileType.XML)
        return 'XML';
    return 'UNKNOWN';
};

exports.FileTypes = FileTypes;

File: objc/objcfind/objcfind_lib/FileTypes.m
#import "config.h"
#import "FileTypes.h"
#import "FileUtil.h"

@interface FileTypes ()


@property NSString *archive;
@property NSString *binary;
@property NSString *code;
@property NSString *text;
@property NSString *unknown;
@property NSString *xml;


@property NSDictionary<NSString*,NSSet<NSString*>*> *fileTypeExtDict;
@property NSDictionary<NSString*,NSSet<NSString*>*> *fileTypeNameDict;

@end

@implementation FileTypes

- (instancetype) init {
    self = [super init];
    if (self) {
        self.archive = [NSString stringWithUTF8String:T_ARCHIVE];
        self.binary = [NSString stringWithUTF8String:T_BINARY];
        self.code = [NSString stringWithUTF8String:T_CODE];
        self.text = [NSString stringWithUTF8String:T_TEXT];
        self.unknown = [NSString stringWithUTF8String:T_UNKNOWN];
        self.xml = [NSString stringWithUTF8String:T_XML];
        NSArray<NSDictionary<NSString*,NSSet<NSString*>*>*> *ftArr = [self fileTypesFromJson];
        self.fileTypeExtDict = ftArr[0];
        self.fileTypeNameDict = ftArr[1];
    }
    return self;
}

- (NSArray<NSDictionary<NSString*,NSSet<NSString*>*>*>*) fileTypesFromJson {
    NSMutableString *fileTypesJsonPath = [NSMutableString stringWithUTF8String:SHAREDPATH];
    [fileTypesJsonPath appendString:@"/filetypes.json"];

    NSMutableDictionary *fileTypeExtDict = [[NSMutableDictionary alloc] init];
    NSMutableDictionary *fileTypeNameDict = [[NSMutableDictionary alloc] init];

    if (![[NSFileManager defaultManager] fileExistsAtPath:fileTypesJsonPath]) {
        return nil;
    }

    NSData *data = [NSData dataWithContentsOfFile:fileTypesJsonPath];

    if (NSClassFromString(@"NSJSONSerialization")) {
        NSError *error = nil;
        id jsonObject = [NSJSONSerialization
                         JSONObjectWithData:data
                         options:0
                         error:&error];

        if (error) { /* JSON was malformed, act appropriately here */ }

        if ([jsonObject isKindOfClass:[NSDictionary class]]) {
            NSArray *fileTypes = jsonObject[@"filetypes"];
            for (NSDictionary *typeDict in fileTypes) {
                NSString *type = typeDict[@"type"];
                NSArray *extensions = typeDict[@"extensions"];
                fileTypeExtDict[type] = [NSSet setWithArray:extensions];
                NSArray *names = typeDict[@"names"];
                fileTypeNameDict[type] = [NSSet setWithArray:names];
            }
        }
    }
    NSMutableArray<NSDictionary<NSString*,NSSet<NSString*>*>*> *fileTypesDictionaries = [[NSMutableArray alloc] init];
    NSDictionary *ftExtDict = [NSDictionary dictionaryWithDictionary:fileTypeExtDict];
    [fileTypesDictionaries addObject:ftExtDict];

    NSDictionary *ftNameDict = [NSDictionary dictionaryWithDictionary:fileTypeNameDict];
    [fileTypesDictionaries addObject:ftNameDict];

    return [NSArray arrayWithArray:fileTypesDictionaries];

}

+ (FileType) fromName:(NSString*)typeName {
    NSString *lname = [typeName lowercaseString];
    if (lname == [NSString stringWithUTF8String:T_TEXT]) {
        return FileTypeText;
    }
    if (lname == [NSString stringWithUTF8String:T_BINARY]) {
        return FileTypeBinary;
    }
    if (lname == [NSString stringWithUTF8String:T_ARCHIVE]) {
        return FileTypeArchive;
    }
    if (lname == [NSString stringWithUTF8String:T_CODE]) {
        return FileTypeCode;
    }
    if (lname == [NSString stringWithUTF8String:T_XML]) {
        return FileTypeXml;
    }
    return FileTypeUnknown;
}

+ (NSString*) toName:(FileType)fileType {
    if (fileType == FileTypeText) {
        return [NSString stringWithUTF8String:T_TEXT];
    }
    if (fileType == FileTypeBinary) {
        return [NSString stringWithUTF8String:T_BINARY];
    }
    if (fileType == FileTypeArchive) {
        return [NSString stringWithUTF8String:T_ARCHIVE];
    }
    if (fileType == FileTypeCode) {
        return [NSString stringWithUTF8String:T_CODE];
    }
    if (fileType == FileTypeXml) {
        return [NSString stringWithUTF8String:T_XML];
    }
    return [NSString stringWithUTF8String:T_UNKNOWN];
}

- (FileType) getFileType:(NSString*)fileName {
    if ([self isCodeFile:fileName]) {
        return FileTypeCode;
    }
    if ([self isXmlFile:fileName]) {
        return FileTypeXml;
    }
    if ([self isTextFile:fileName]) {
        return FileTypeText;
    }
    if ([self isBinaryFile:fileName]) {
        return FileTypeBinary;
    }
    if ([self isArchiveFile:fileName]) {
        return FileTypeArchive;
    }
    return FileTypeUnknown;
}

- (BOOL) isFileOfType:(NSString*)fileName type:(NSString*)typeName {
    if ([self.fileTypeNameDict[typeName] containsObject:fileName]) {
        return true;
    }
    NSString *ext = [FileUtil getExtension:fileName];
    return self.fileTypeExtDict[typeName] != nil &&
    [self.fileTypeExtDict[typeName] containsObject:ext];
}

- (BOOL) isArchiveFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.archive];
}

- (BOOL) isBinaryFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.binary];
}

- (BOOL) isCodeFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.code];
}

- (BOOL) isTextFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.text];
}

- (BOOL) isXmlFile:(NSString*)fileName {
    return [self isFileOfType:fileName type:self.xml];
}

- (BOOL) isFindableFile:(NSString*)fileName {
    return [self getFileType:fileName] != FileTypeUnknown;
}

- (BOOL) isUnknownFile:(NSString*)fileName {
    return [self getFileType:fileName] == FileTypeUnknown;
}

@end
